### PR TITLE
use a watch to convey cache changes outside the cache

### PR DESCRIPTION
**Pull request by flyingmachine**
There are a lot of things that I think could be different about this PR, and it might be better to take [a different approach entirely](https://github.com/metabase/metabase/issues/10630#issuecomment-522824933) but I wanted to create it anyway :)

At the very least it can serve as a good anchor for discussion to help me learn more about how things are organized, conventions to follow, stuff like that.

### CODE REVIEW

File: src/metabase/models/setting.clj [hunk 1]
```diff
@@ -98,6 +104,18 @@
...
 
+(defn- call-on-change
+  "Cache watcher that applies `:on-change` callback for all settings that have changed."
+  [_key _ref old new]
+  (let [rs @registered-settings]
+    (doseq [changed-setting (-> (data/diff old new)
+                                (second)
+                                (keys))]
+      (when-let [on-change (get-in rs [(keyword changed-setting) :on-change])]
+        (on-change (clojure.core/get old changed-setting) (clojure.core/get new changed-setting))))))
```
Comment (camsaul) [inline src/metabase/models/setting.clj]: does it make sense to call `on-change` if the `old` and `new` values are the same?

File: test/metabase/models/setting/cache_test.clj [hunk 2]
```diff
@@ -122,3 +124,28 @@
     ;; detect a cache out-of-date situation and flush the cache as appropriate, giving us the updated value when we
     ;; call! :wow:
     (setting-test/toucan-name)))
+
+;; sets site locale setting
+(expect
+  "fr"
+  (let [original-locale (java.util.Locale/getDefault)
```
Comment (camsaul) [inline test/metabase/models/setting/cache_test.clj]: Can you put the code that restores the default locale in a `try-finally` block? That way if the tests throw an Exception we don't end up running other tests in French which could lead to hard-to-debug test failures

Comment (flyingmachine) [reply test/metabase/models/setting/cache_test.clj]: sure thing!

File: src/metabase/models/setting.clj (hunk 1 above)
Comment (flyingmachine) [reply src/metabase/models/setting.clj]: `on-change` will only get called if the values are different

**Comment by flyingmachine** [comment]
@camsaul this is ready for another look

**Review by camsaul — APPROVED** [review]

**Pull request merged**